Content.Server/EntityEffects/Effects/ActivateRandom.cs

```diff
@@ -0,0 +1,166 @@
+                            : null)
+                        .Where(x => x is not null))
+                + ":\n"
+                + string.Join("\n", case_.Effects.ToList()
+                    .Select(effect =>
+                    {
+                        var desc = effect.GuidebookEffectDescription(prototype, entSys);
+                        return desc is null ? null : "  " + desc;
+                    })
+                    .Where(x => x is not null))))
+                + "\n"));
+    }
+
+    public override void Effect(EntityEffectBaseArgs baseArgs)
+    {
+         // ActivateRandom works by activating other reagents, it needs access to extended information
+        if (baseArgs is not EntityEffectReagentArgs)
+            return;
+        EntityEffectReagentArgs args = (EntityEffectReagentArgs) baseArgs;
+            
+        if (args.Source is null)
+            return; // ActivateRandom activates chemicals from the args.Source.
+            // If source is not present, ActivateRandom has nothing to activate.
+
+        // Only one random set of chemicals is activated. The chance to activate
+        // some speicific set is based on how many of these chemicals (with respect to factor)
+        // are present in the solution.
+        // We use chance and chanceSum to calculate proprotions of chemical sets (with respect to factor)
+        // present in the solution.
+        float[] chances = new float[Cases.Length];
+        float chancesSum = 0f;
+
+        for (int i = 0; i < Cases.Length; i++)
+        {
+            var _case = Cases[i];
+            foreach (var activate in _case.Activates)
+            {
+                if (args.Source.TryGetReagent(activate.Reagent, out var quantity))
+                {
+                    var impact = quantity.Quantity.Float() * activate.Factor.Float();
+                    chances[i] += impact;
+                    chancesSum += impact;
+                }
+            }
+        }
+
+        // Next, select a "random" set. Sets of higher propotions are prioritized.
+        int index = 0;
+        IRobustRandom _random = IoCManager.Resolve<IRobustRandom>();
+        {
+            float randval = _random.NextFloat(0f, chancesSum);
+            for (int i = 0; i < Cases.Length && randval > 0f; i++) {
+                index = i;
+                randval -= chances[i];
+            }
+        }
+
+        // When some set is activated, ActivateRandom removes reagents of the set marked as "consumed".
+        // When the effects of the set are run, the amount of *consumed* reagents as passed to them.
+        // This is useful, for example, in conjuction with HealthChange's scaleByQuantity
+        FixedPoint2 consumed = 0f;
+        foreach (var activate in Cases[index].Activates) {
+            if (activate.Consume && args.Source.TryGetReagent(activate.Reagent, out var quantity)) {
+                consumed += quantity.Quantity;
```

Comment:
as above on newline, also it should be FixedPoint2.Empty or FixedPoint2.Zero I forget which.